Use session to store shortUrl visits, avoiding repeated transition page redirects

yong-jie opened the pull request:
## Problem

We need a way to not show the transition page multiple times for the same shortUrl.

## Solution

Use session to store the shortUrl, and read it before deciding to redirect directly or show transition page. Sessions are much easier to work with compared to cookies, as we can see how difficult parsing the cookie was with GA.

yong-jie commented:
PR has been updated to use cookie-session. A sliding window is also used to prune old shortUrl visits.

liangyuanruo commented on src/server/api/redirect.ts:
```diff
@@ -171,18 +180,52 @@ export default async function redirect(req: Express.Request, res: Express.Respon
 
     // Redirect immediately if a crawler is visiting the site
-    if (isCrawler(req.headers["user-agent"] || '')) {
+    if (isCrawler(req.headers['user-agent'] || '')) {
```
Apparently this is the canonical way to obtain the header according to [Express docs](https://expressjs.com/en/api.html)
```suggestion
    if (isCrawler(req.get('user-agent') || '')) {
```

liangyuanruo commented on src/server/api/redirect.ts:
```diff
@@ -171,18 +180,52 @@ export default async function redirect(req: Express.Request, res: Express.Respon
+    }
+
+    // Reassure typescript that the session object exists.
+    if (!req.session) {
```
we should simply use `req.session!.property`, instead of modifying our program's behavior and writing code for the compiler.

liangyuanruo commented on src/server/api/redirect.ts:
```diff
@@ -171,18 +180,52 @@ export default async function redirect(req: Express.Request, res: Express.Respon
+      }
+
+      // Avoid exceeding cookie size limit by pruning
+      // old entries.
```
I have a concern with limiting by an arbitrary number of short links, which is susceptible to variability in the length of each short link, when we are able to compute exactly the number of bytes that will be taken up.

As the characters that make up the short links and epoch are a subset of ASCII, the size in bytes is also the length of the string. Can we use that instead to guarantee keeping under the 4k limit?

We might also leave some buffer and use just e.g. 3k.

FInally, could this be extracted into a separate function?

liangyuanruo commented on src/server/api/redirect.ts:
```diff
@@ -171,18 +180,52 @@ export default async function redirect(req: Express.Request, res: Express.Respon
     }
+
+    // User has visited this shortlink before.
+    // Update LRU cache and redirect.
```
how about writing an LRU cache class with a `maxSize` parameter in bytes? This would help to manage complexity in the codebase. Apart from regular get/set functions, you could easily extend to add a `toObject()` function that allows reassignment back to `req.session`.

yong-jie commented:
Rebased the branch on top of kyler's recently-merged PR. Implemented LRU cache class that abstracts away all the storage, pruning, get/set logic. The LRU cache now tracks storage size based on the length of the json-string, with a conservative max size of 2000 bytes.

liangyuanruo commented on src/server/util/lrucache.ts:
```diff
@@ -0,0 +1,98 @@
+
+const MAX_SIZE_BYTES = 2000
+
+interface CookieData {
```
let's strive to make this LRUCache class standalone as an independent data structure - it currently violates the _Single Responsibility Principle_ because it knows a lot about the format of the cookies, mapping epoch to short links etc.

we could also use _types_ such as [Record<K, T>](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt) instead of interfaces here.

liangyuanruo commented on src/server/util/lrucache.ts:
```diff
@@ -0,0 +1,98 @@
+
+  lookupTable: EpochToShortUrlMapping = {}
+
+  timestamps: number[] = []
```
checkout [this implementation](https://medium.com/sparkles-blog/a-simple-lru-cache-in-typescript-cba0d9807c40) of an LRUCache. it relies on the fact that Javascript `Map`s are actually ordered, so the implementation can be simplified significantly.

liangyuanruo replied:
actually the correct data structure here should be a Priority Queue.

when a redirect request arrives, the server simply iterates through the existing key-value pairs and insert all records, which will be automatically sorted, then upserts the existing short link/epoch (upsert is achieved by removing and reinserting).

to build up the new cookie, simply take from the PQ until iterator is exhausted or the size limit is hit.

does this make sense? if it does perhaps it will be easier to look for an existing implementation instead of rolling our own.

liangyuanruo replied:
Can look at Priority Queues from [google closure](https://google.github.io/closure-library/api/goog.structs.PriorityQueue.html) or [BucketJS](http://mauriciosantos.github.io/Buckets-JS/symbols/buckets.PriorityQueue.html)

liangyuanruo replied:
thinking yet again - maybe we're over-engineering here. looking at this pattern:

```javascript
const cache = new LRUCache(req.session!.visits)
cache.updateEntry(shortUrl)
req.session!.visits = cache.getData()
```

we simply need a reducer function

```javascript
req.session!.visits = updateCookies(req.session!.visits || {}, shortUrl)
```

where

```javascript
type updateCookies = (visits: Record<string, number>, shortUrl: string) => Record<string, number>
```
